Watcher handle graceful onclose

NiklasJonsson6 opened the pull request:
We mistakenly did not implement any behavior for when our watch for kubernetes endpoints closed gracefully, since it [has a default implementation that only logs a message on the debug level.](https://github.com/fabric8io/kubernetes-client/blob/main/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Watcher.java#L44)

- [x] Log on DEBUG-level for the kubernetes client.

- [x] Handle graful watcher close the same way we handle exception close (log and exit the application with code 11).

NiklasJonsson6 committed:
exit quarkus on graceful watch onClose

the same way as we do when the watch closes exceptionally. The graceful onClose is default-implemented with a debug log message so we had mistakenly not overridden it.

NiklasJonsson6 committed:
log debug level for kubernetes client

We had no sign of graceful watcher onClose since it only logged a debug message in its default implementation

solsson commented on src/main/resources/application.yaml:
```diff
@@ -50,6 +50,8 @@ quarkus:
       "org.apache.kafka.clients.Metadata":
         level: DEBUG
+      "io.fabric8.kubernetes.client":
+        level: DEBUG
```
We should use an env here. This is great for troubleshooting stuff like ECONNRESET but we should probably default to info.

solsson replied:
Let's do a single-arch build with DEBUG, so we don't have to update lots of yaml.

solsson committed:
Quarkus 3.6.0->3.6.3

solsson committed:
io.fabric8.kubernetes.client.vertx only logs errors

solsson reviewed (approved):
Do we want to test any of these, from [client reference](https://quarkus.io/guides/kubernetes-client#configuration-reference)?

<img width="961" alt="Screenshot 2023-12-18 at 16 43 24" src="https://github.com/Yolean/kafka-keyvalue/assets/144945/f355e171-9060-4edf-a047-031ad96ab35c">

solsson committed:
Counts both success and failure so we can get an error rate

solsson committed:
Uses the same prefix as our other onupdate related metrics

solsson committed:
Changed my mind: keep the prefix, means actual watch target update

while "onupdate" refers to the mechanism of handling key-value updates

This reverts commit 440a8cd4f871e2842bf8ef08b8f7c15ebfbbffb3.

solsson committed:
Fixes test with non-static init

NiklasJonsson6 commented:
> Do we want to test any of these, from [client reference](https://quarkus.io/guides/kubernetes-client#configuration-reference)?
> 
> <img alt="Screenshot 2023-12-18 at 16 43 24" width="961" src="https://private-user-images.githubusercontent.com/144945/291339121-f355e171-9060-4edf-a047-031ad96ab35c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5NzE1NTIsIm5iZiI6MTcwMjk3MTI1MiwicGF0aCI6Ii8xNDQ5NDUvMjkxMzM5MTIxLWYzNTVlMTcxLTkwNjAtNGVkZi1hMDQ3LTAzMWFkOTZhYjM1Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxOVQwNzM0MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZGZmZDAzNjBjZmI3MTZiNzI3YWQyNTIyZGM0ZTkwMWQ3YmI0ODE0NTE4YzcxMjQ1NzM1YWFhN2RiM2M0NTA2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vQhJBmbBd9ydOsFrqg3XMOctn4WmFWedMV1R8ci0sXk">

I think no, since the default reconnect behavior seems quite agressive. The issue we had was (most likely) a graceful close of the watch according to the client, so in that case I would assume no reconnect is done. Otherwise, an unlimited of retry with a 1 second interval (the default reconnect behavior) would surely have succeeded eventually.

solsson committed:
Adds reconnect logging

solsson commented:
[Watching](https://github.com/fabric8io/kubernetes-client/blob/v6.9.2/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/AbstractWatchManager.java#L314) is never logged, despite debug level. Probably a configuration error on our side. [WebSocket](https://github.com/fabric8io/kubernetes-client/blob/v6.9.2/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatcherWebSocketListener.java#L40) is also never logged.

solsson committed:
should only be needed for trace level, but why don't we get logs?

solsson committed:
Quarkus 3.6.4

solsson committed:
Don't cache jvm builds, they're cheap

solsson committed:
Avoids exit, allows the client lib to retry

solsson committed:
Reconnect on close, as we're not using an informer

solsson commented:
Evaluating `docker.io/yolean/kafka-keyvalue:509399df255d03f7fe4fb63a0b5c653b7ddf8aab@sha256:6eec474b2a58870bc404523ddabd92cbdefbe78bcb29e85891042f7721bd7aa5` with GKE 1.27 API-server. It seems to recover from "too old resource version" and keep endpoints unique and up-to-date.

I haven't added test coverage because I think we might want to switch to an informer (or are they always watching all namespaces?) in which case none of these hacks will be needed, at least not in their current form.

solsson commented:
I still don't know how API-server restarts behave now. We do however have debug logging for the websocket connections since 252f46d7eb236ffbcb9d704dd4ededa381fe70d7.
 .
https://github.com/fabric8io/kubernetes-client/issues/5189

solsson commented:
https://github.com/fabric8io/kubernetes-client/issues/5372#issuecomment-1665508031

> This is handled automatically by Informers, but not for Watches. Bookmarks will be used by default, which for newer api servers make this exception much more infrequent.

NiklasJonsson6 committed:
Quarkus 3.6.7

NiklasJonsson6 committed:
integration test for endpoints watch reconnect behavior

using testcontainers with k3s

NiklasJonsson6 committed:
EndpointsWatcher retries on connection closed

solsson commented on src/main/java/se/yolean/kafka/keyvalue/kubernetes/EndpointsWatcher.java:
```diff
@@ -131,7 +139,32 @@ void handleEvent(Action action, Endpoints resource) {
+      vwatch.close();
+    }
+
+    double jitter = Math.random() * 0.2 + 1;
```
I think the range and unit should be clearly stated here, because it affects the load on the API-server after lots of pods have disconnected at the same time.

solsson reviewed (changes requested):
My interpretation is that we restart watches all the time. Correct?

NiklasJonsson6 commented:
After the latest changes, the integration test validates that we reconnect our watch after being disconnected. Running the test confirms that we actually get onClose-events after reconnect fails. This might be due to the fact that we're running a newer version now, and issues like these are fixed https://github.com/fabric8io/kubernetes-client/issues/5189.

If the integration test failed to reproduce the issue, we're back to square one...

NiklasJonsson6 committed:
configure min- and max-retry delay

and select a random value between to use

NiklasJonsson6 committed:
plain watch replaced with informer

the informer handles reconnects internally, which makes in unneccesary for us to write our own logic for that. Also, the replaced watch reconnect behaviour we wrote ourselves was not perfect and regularly re-watched without a need to do so

NiklasJonsson6 committed:
add synchornized to handleEvent method

we modify two collections in sequence and the method is invoked by random threadpool-threads

solsson commented on src/main/resources/application.yaml:
```diff
@@ -1,7 +1,7 @@
 kkv:
+  namespace: ${NAMESPACE}
```
By convention we use POD_NAMESPACE when depending on the [downward API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/) for envs.

Alternatively I think we should use `WATCH_NAMESPACE` or something like `TARGET_NAMESPACE` if we think that watching a different namespace is a reasonable use case.

NiklasJonsson6 replied:
I'll move `namespace` and `resync-period` to `kkv.target.service` (next to service name and port), since they are directly related to this. 

I think that `TARGET_NAMESPACE`, or even more descriptive `TARGET_SERVICE_NAMESPACE` is a good idea. Even though we always run kkv in the same namespace as the service, I really don't think there are any technical restrictions to allowing other namespaces as well.

NiklasJonsson6 committed:
config property and metrics fix

NiklasJonsson6 committed:
quarkus 3.6.8

NiklasJonsson6 committed:
require targetServiceNamespace if informer is enabled

solsson committed:
Builder with latest Mandrel and JRE runtime with java 21.0.2
